internal/expr/bindtypes.go

```diff
@@ -49,64 +49,93 @@ func (pe *ParsedExpr) BindTypes(args ...any) (tbe *TypeBoundExpr, err error) {
+		if err != nil {
+			return nil, err
+		}
 
+		if toe, ok := te.(*typedOutputExpr); ok {
 			for _, oc := range toe.outputColumns {
 				if ok := outputUsed[oc.output]; ok {
 					return nil, fmt.Errorf("%s appears more than once in output expressions", oc.output.String())
 				}
 				outputUsed[oc.output] = true
 			}
-			te = toe
-		case *bypass:
-			te = e
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
 		}
 		typedExprs = append(typedExprs, te)
 	}
 
 	return &typedExprs, nil
 }
 
-// bindInputTypes binds the input expression to a query type and returns a typed
-// input expression.
-func bindInputTypes(e *inputExpr, argInfo typeinfo.ArgInfo) (te *typedInputExpr, err error) {
-	defer func() {
-		if err != nil {
-			err = fmt.Errorf("input expression: %s: %s", err, e.raw)
-		}
-	}()
+// expression represents a parsed node of the SQLair query's AST.
+type expression interface {
+	// String returns a text representation for debugging and testing purposes.
+	String() string
 
-	var input typeinfo.Input
-	switch a := e.sourceType.(type) {
-	case memberAccessor:
-		input, err = argInfo.InputMember(a.typeName, a.memberName)
-		if err != nil {
-			return nil, err
-		}
-	case sliceAccessor:
-		return nil, fmt.Errorf("slice support not implemented")
+	// bindTypes binds the types to the expression to generate either a
+	// *typedInputExpr or *typedOutputExpr.
+	bindTypes(typeinfo.ArgInfo) (any, error)
+}
+
+// memberInputExpr is an input expression of the form "$Type.member" which
+// represents a query parameter contained in a member of a type.
+type memberInputExpr struct {
+	raw string
+	ma  memberAccessor
+}
+
+// String returns a text representation for debugging and testing purposes.
+func (e *memberInputExpr) String() string {
+	return fmt.Sprintf("Input[%+v]", e.ma)
+}
+
+// bindTypes generates a *typedInputExpr containing type information about the
+// Go object and its member.
+func (e *memberInputExpr) bindTypes(argInfo typeinfo.ArgInfo) (any, error) {
+	input, err := argInfo.InputMember(e.ma.typeName, e.ma.memberName)
+	if err != nil {
+		return nil, fmt.Errorf("input expression: %s: %s", err, e.raw)
 	}
 	return &typedInputExpr{input}, nil
 }
 
-// bindOutputTypes binds the output expression to concrete types. It then checks the
-// expression valid with respect to its bound types and generates a typed output
-// expression.
-func bindOutputTypes(e *outputExpr, argInfo typeinfo.ArgInfo) (te *typedOutputExpr, err error) {
+// sliceInputExpr is an input expression of the form "$S[:]" that represents a
+// slice of query parameters.
+type sliceInputExpr struct {
+	raw           string
+	sliceTypeName string
+}
+
+// String returns a text representation for debugging and testing purposes.
+func (e *sliceInputExpr) String() string {
+	return fmt.Sprintf("Input[%s[:]]", e.sliceTypeName)
+}
+
+// bindTypes generates a *typedInputExpr containing type information about the
```

Comment:
In my opinion, all of these comments about the output type should be in the types themselves.

Reply:
In my opinion they should really be implicit in the type, but since we need to implement the interface that has an `any` its necessary to specify it. I would rather not put them in the function because they define the result of the function, not how it works.

It would make sense to remove the output type from the function name if we replace the `any` in the `bindTypes` function in the interface with an interface such as `typedExpr`. I did try doing that but it was hard to make a compelling case to myself that it added anything more than it did in PR #127 when we decided not to add it.

Reply:
I was advocating for your second paragraph. If you tried it and it does not fit then let's leave it like in this PR.

Reply:
Yea, I felt that Joe's comment from before still stands so there is no justification to change it.